src/OutputWriters/netcdf_output_writer.jl

```diff
@@ -1,135 +1,716 @@
+        xᶜᵃᵃ_name = dim_name_generator("x", grid, c, nothing, nothing, Val(:x))
+
+        xᶠᵃᵃ_data = collect_dim(grid.xᶠᵃᵃ, f, TX(), Nx, Hx, indices[1], with_halos)
+        xᶜᵃᵃ_data = collect_dim(grid.xᶜᵃᵃ, c, TX(), Nx, Hx, indices[1], with_halos)
+
+        dims[xᶠᵃᵃ_name] = xᶠᵃᵃ_data
+        dims[xᶜᵃᵃ_name] = xᶜᵃᵃ_data
+    end
+
+    if TY != Flat
+        yᵃᶠᵃ_name = dim_name_generator("y", grid, nothing, f, nothing, Val(:y))
+        yᵃᶜᵃ_name = dim_name_generator("y", grid, nothing, c, nothing, Val(:y))
+
+        yᵃᶠᵃ_data = collect_dim(grid.yᵃᶠᵃ, f, TY(), Ny, Hy, indices[2], with_halos)
+        yᵃᶜᵃ_data = collect_dim(grid.yᵃᶜᵃ, c, TY(), Ny, Hy, indices[2], with_halos)
+
+        dims[yᵃᶠᵃ_name] = yᵃᶠᵃ_data
+        dims[yᵃᶜᵃ_name] = yᵃᶜᵃ_data
+    end
+
+    if TZ != Flat
+        vertical_dims = gather_vertical_dimensions(grid.z, TZ, Nz, Hz, indices[3], with_halos, dim_name_generator)
+        dims = merge(dims, vertical_dims)
+    end
+
+    maybe_add_particle_dims!(dims, outputs)
+
+    return dims
+end
+
+function gather_dimensions(outputs, grid::LatitudeLongitudeGrid, indices, with_halos, dim_name_generator)
+    TΛ, TΦ, TZ = topology(grid)
+    Nλ, Nφ, Nz = size(grid)
+    Hλ, Hφ, Hz = halo_size(grid)
+
+    dims = Dict()
+
+    if TΛ != Flat
+        λᶠᵃᵃ_name = dim_name_generator("longitude", grid, f, nothing, nothing, Val(:x))
+        λᶜᵃᵃ_name = dim_name_generator("longitude", grid, c, nothing, nothing, Val(:x))
+
+        λᶠᵃᵃ_data = collect_dim(grid.λᶠᵃᵃ, f, TΛ(), Nλ, Hλ, indices[1], with_halos)
+        λᶜᵃᵃ_data = collect_dim(grid.λᶜᵃᵃ, c, TΛ(), Nλ, Hλ, indices[1], with_halos)
+
+        dims[λᶠᵃᵃ_name] = λᶠᵃᵃ_data
+        dims[λᶜᵃᵃ_name] = λᶜᵃᵃ_data
+    end
+
+    if TΦ != Flat
+        φᵃᶠᵃ_name = dim_name_generator("latitude", grid, nothing, f, nothing, Val(:y))
+        φᵃᶜᵃ_name = dim_name_generator("latitude", grid, nothing, c, nothing, Val(:y))
+
+        φᵃᶠᵃ_data = collect_dim(grid.φᵃᶠᵃ, f, TΦ(), Nφ, Hφ, indices[2], with_halos)
+        φᵃᶜᵃ_data = collect_dim(grid.φᵃᶜᵃ, c, TΦ(), Nφ, Hφ, indices[2], with_halos)
+
+        dims[φᵃᶠᵃ_name] = φᵃᶠᵃ_data
+        dims[φᵃᶜᵃ_name] = φᵃᶜᵃ_data
+    end
+
+    if TZ != Flat
+        vertical_dims = gather_vertical_dimensions(grid.z, TZ, Nz, Hz, indices[3], with_halos, dim_name_generator)
+        dims = merge(dims, vertical_dims)
+    end
+
+    maybe_add_particle_dims!(dims, outputs)
+
+    return dims
+end
+
+gather_dimensions(outputs, grid::ImmersedBoundaryGrid, args...) =
+    gather_dimensions(outputs, grid.underlying_grid, args...)
+
+#####
+##### Gathering of grid metrics
+#####
+
+function gather_grid_metrics(grid::RectilinearGrid, indices, dim_name_generator)
+    TX, TY, TZ = topology(grid)
+
+    metrics = Dict()
+
+    if TX != Flat
+        Δxᶠᵃᵃ_name = dim_name_generator("dx", grid, f, nothing, nothing, Val(:x))
+        Δxᶜᵃᵃ_name = dim_name_generator("dx", grid, c, nothing, nothing, Val(:x))
```

Comment:
It seems that `dx` is hardcoded here. Should we allow for more flexibility here too since this is one of the motivations behind the PR?

Reply:
Ah so I suppose `dx` is the "canonical" name, but the `dim_name_generator` can be used to override this name!

For example, a custom `dim_name_generator` can be passed to the `NetCDFOutputWriter` constructor to change `dx` to `Δx` or `delta_x` before passing it back to `Oceananigans.OutputWriters.default_dim_name`.

This is totally not obvious from the code (and the docstring) so I should add a test for this and probably another docstring example. It's how I was thinking we can satisfy desired naming conventions.

But happy to switch to another better design!

Reply:
Yeah I figured that somewhere down in between this and the last comment haha. There's a comment somewhere about it. I think the way you're doing is great!